sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/DistinctTest.java

```diff
@@ -118,11 +131,122 @@ public void testDistinctWithRepresentativeValue() {
+            TimestampedValue.of("k3", base.plus(Duration.standardSeconds(50))),
+            TimestampedValue.of("k4", base.plus(Duration.standardSeconds(60))),
+            TimestampedValue.of("k5", base.plus(Duration.standardSeconds(70))),
+            TimestampedValue.of("k6", base.plus(Duration.standardSeconds(80))))
+        .advanceWatermarkToInfinity();
+
+    PCollection<String> distinctValues = windowedDistinctPipeline
+        .apply(values)
+        .apply(Window.<String>into(FixedWindows.of(Duration.standardSeconds(30))))
+        .apply(Distinct.<String>create());
+    PAssert.that(distinctValues)
+        .inWindow(new IntervalWindow(base, base.plus(Duration.standardSeconds(30))))
+        .containsInAnyOrder("k1", "k2", "k3");
+    PAssert.that(distinctValues)
+        .inWindow(new IntervalWindow(base.plus(
+            Duration.standardSeconds(30)), base.plus(Duration.standardSeconds(60))))
+        .containsInAnyOrder("k1", "k2", "k3");
+    PAssert.that(distinctValues)
+        .inWindow(new IntervalWindow(base.plus(
+            Duration.standardSeconds(60)), base.plus(Duration.standardSeconds(90))))
+        .containsInAnyOrder("k4", "k5", "k6");
+    windowedDistinctPipeline.run();
+  }
+
+  @Rule
+  public TestPipeline triggeredDistinctPipeline = TestPipeline.create();
+
+  @Test
+  @Category({ValidatesRunner.class, UsesTestStream.class})
+  public void testTriggeredDistinct() {
+    Instant base = new Instant(0);
+    TestStream<String> values = TestStream.create(StringUtf8Coder.of())
+        .advanceWatermarkTo(base)
+        .addElements(
+            TimestampedValue.of("k1", base),
+            TimestampedValue.of("k2", base.plus(Duration.standardSeconds(10))),
+            TimestampedValue.of("k3", base.plus(Duration.standardSeconds(20))))
+        .advanceProcessingTime(Duration.standardMinutes(1))
+        .addElements(
+            TimestampedValue.of("k1", base.plus(Duration.standardSeconds(30))),
+            TimestampedValue.of("k2", base.plus(Duration.standardSeconds(40))),
+            TimestampedValue.of("k3", base.plus(Duration.standardSeconds(50))))
+        .advanceWatermarkToInfinity();
+
+    PCollection<String> distinctValues = triggeredDistinctPipeline
+        .apply(values)
+        .apply(Window.<String>into(FixedWindows.of(Duration.standardMinutes(1)))
+            .triggering(Repeatedly.forever(
+                AfterProcessingTime.pastFirstElementInPane().plusDelayOf(
+                    Duration.standardSeconds(30))))
+            .withAllowedLateness(Duration.ZERO)
+            .accumulatingFiredPanes())
+        .apply(Distinct.<String>create());
+    PAssert.that(distinctValues).containsInAnyOrder("k1", "k2", "k3");
+    triggeredDistinctPipeline.run();
+  }
+
+  @Rule
+  public TestPipeline triggeredDistinctRepresentativePipeline = TestPipeline.create();
+
+  @Test
+  @Category({ValidatesRunner.class, UsesTestStream.class})
```

Comment:
I think these are `NeedsRunner` rather than `ValidatesRunner`. They are not testing the runner, but the transform.